### ISSUE TITLE
ServiceBusTrigger is not triggered after function goes asleep

### ISSUE THREAD

**Issue by fdario-gcm** [Azure/azure-functions-host#4366]
We have a Function App V2 that has TimeTriggers, HttpTriggers and ServiceBusTriggers, running on Consumption Plan.
Hitting the function URL or doing any HTTP request "wakes up" the function, and once the function is up and running, every message in a Service Bus queue gets picked up and processed correctly. After some idle minutes, the function, as expected, goes asleep, and it simply doesn't trigger anymore when new messages are sent to the queue.

#### Investigative information

- Timestamp: 2019-04-29T17:48:22.1330372Z
- Function App version (1.0 or 2.0): 2.0 (2.0.12427.0)
- Invocation ID: 5df55e1a-9629-4824-b9d0-b8d83fb0f6a0
- Operation ID: 131a04383fc0174da2794efe0e14ccc6
- Region: East US

#### Repro steps

1. Hit the function URL in the browser;
2. Send a message to a Service Bus queue;
3. The message gets picked up as expected;
4. Wait approximately 10 minutes for the function to go asleep;
5. Send another message to a Service Bus queue;
6. The message stays on the queue until any of the following events happen:
6.1 A Time Trigger runs;
6.2 An HttpTrigger receives a request;
6.3 The function URL is opened in a browser;
6.4. Eventually it gets picked up without any of the events above happening, but it's usually in the next day, close to 9AM. Might be related to a TimeTrigger we have running every day at 9AM, because the function should have to start moments before the TimeTrigger run time, so it would be able to pick up the remaining messages from the queue.

#### Expected behavior

Whenever a message is sent to a Service Bus Queue, the function should pick it up. If it's dormant, it should wake up and pick it up.

#### Actual behavior

As soon as a function V2 in Consumption Plan goes asleep, no more messages are picked up from a Service Bus queue or topic subscription

- At 2019-04-29 17:48:22.133Z, a message was last picked up from a Service Bus queue, because the function was awake;
- At 2019-04-29 20:32:08:226Z a message was sent to the same Service Bus queue;
- It remained in the queue until 2019-04-30T08:58:50.269Z, when it got picked up, approximately 1 minute before our daily Time Trigger runs (2019-04-30 09:00:00.006Z).

Last time the function picked up a message before going asleep:
Timestamp: 2019-04-29T17:48:22.1330372Z
InvocationId: 5df55e1a-9629-4824-b9d0-b8d83fb0f6a0
Duration: 1,664.597
OperationId: 131a04383fc0174da2794efe0e14ccc6

Last time the function ran and picked up a message, more than 12 hours after the last message was put in the queue, and moments before our daily TimeTrigger runs:
Timestamp: 2019-04-30T08:58:50.2695977Z
InvocationId: 48afc4cd-824d-4bb6-97ad-293a86e6f6fd
Duration: 4,121.5649
OperationId: 73e4e02ee116ff46a2205e2e21b54f2f

#### Known workarounds

Hitting the function URL in the browser, doing any HTTP request to a HttpTrigger or having a TimeTrigger running every 5min wakes up the function and the messages are picked up as expected.

We had the function running for a couple days with a TimeTrigger running every 5 minutes and we didn't have any issues, all messages were picked up seconds after they were put in the queue.
Once we disabled that, the issue started to happen again.

#### Related information 

* Programming language used: C#
* Precompiled function

**Comment by fabiocav** [comment]
@SatishRanjan can you please help with initial ivestigation and triage? Looks like triggers were not correctly synchronized or credentials are not usable from the controler.

**Comment by jnevins-gcm** [comment]
@fabiocav

This is consistently reproducible with any deployment - we are using the kudu zipdeploy api. 

We have tried adding Invoke-ResourceAction after deployment to explicitly resync triggers after deployment without any effect. 

As for the credentials - please note that the function DOES receive messages successfully before it falls asleep, so our credentials to connect to ASB are correct. 

If it makes any different, our function only specifies queue name in the ServiceBusTrigger - we rely on setting ConnectionStrings:ServiceBus to use as the default connection string.

**Comment by jnevins-gcm** [comment]
@SatishRanjan
@fabiocv

Any update here? This is a production issue for us.

**Comment by jeffhollan** [comment]
Ah yes connectionString:ServiceBus wouldn’t be supported. Any reason you are using that connection string instead of an app setting?'

**Comment by jnevins-gcm** [comment]
I don’t believe AzureWebJobsServiceBus was recognized when we started using v2

**Comment by jnevins-gcm** [comment]
@jeffhollan

We can switch it. I think AzureWebJobsServiceBus as an app setting wasn’t recognized by web jobs as the default service bus connection string to use when we started using v2

**Comment by jnevins-gcm** [comment]
@jeffhollan

we tried switching to AzureWebJobsServiceBus and found that the bug that made us choose ConnectionStrings:ServiceBus still exists

If we don't specify the Connection property on the ServiceBusTrigger, it doesn't find the AzureWebJobsServiceBus setting and use it as the default. We now need to explicitly specify the Connection name in every attribute or explicitly use the class attribute ServiceBusAccount("AzureWebJobsServiceBus")  or we get an error at startup:

https://github.com/Azure/azure-webjobs-sdk/blob/2e374a2c58fd89232945e997141f199723dc04db/src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusAccount.cs#L44

This is because the webjobs sdk still will only use ConnectionStrings:ServiceBus as a default to look for:

https://github.com/Azure/azure-webjobs-sdk/blob/eeb0744bf23d2ce025062a3d33e07ad58a605b8f/src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Config/ServiceBusWebJobsBuilderExtensions.cs#L43

which uses

https://github.com/Azure/azure-webjobs-sdk/blob/eeb0744bf23d2ce025062a3d33e07ad58a605b8f/src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Constants.cs#L27

it's not a big deal and we will add ServiceBusAccount("AzureWebJobsServiceBus") to our classes, but may be worth addressing the bug in webjobs sdk or at least throw an error and not support any default value

**Comment by jnevins-gcm** [comment]
Thanks @jeffhollan

Please see the last part of my post about how we set the connection string. We will give it a try setting explicitly, but if that’s the problem, it sounds like a bug in the scale controller looking at the default connection string name.

**Comment by jeffhollan** [comment]
I believe instead of defining JUST the connection string “ServiceBus” you can create only a AzureWebJobsServiceBus app setting. That way you don’t have to be explicit in your bindings or duplicate

**Comment by jnevins-gcm** [comment]
If you just do that, you get an error from the host on startup here

https://github.com/Azure/azure-webjobs-sdk/blob/2e374a2c58fd89232945e997141f199723dc04db/src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/ServiceBusAccount.cs#L44

**Comment by alrod** [comment]
@jnevins-gc, sorry for late reply. Are you still facing the issue?

**Comment by jnevins-gcm** [comment]
The issue remains exactly as described. 

Long story short - there is no way to use a default connection string and we need to explicitly specify the connection string in a ServiceBusAccount attribute in every class.
